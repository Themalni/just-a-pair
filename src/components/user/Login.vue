<template>
  <div class="modal fade" id="userLoginModal" tabindex="-1" role="dialog" aria-hidden="true">
    <div class="modal-dialog" role="document">
      <div class="modal-content">
        <div class="modal-header text-center">
          <ul class="nav nav-tabs" id="loginTab" role="tablist">
            <li class="nav-item">
              <a
                class="nav-link active mr-1"
                id="login-tab"
                data-toggle="tab"
                href="#login"
                role="tab"
                aria-controls="login"
                aria-selected="true"
              >Login</a>
            </li>
            <li class="nav-item">
              <a
                class="nav-link ml-1"
                id="register-tab"
                data-toggle="tab"
                href="#register"
                role="tab"
                aria-controls="register"
                aria-selected="false"
              >Register</a>
            </li>
          </ul>
        </div>
        <div class="modal-body mx-3">
          <div class="tab-content" id="myTabContent">
            <!-- Login -->
            <div
              class="tab-pane fade show active"
              id="login"
              role="tabpanel"
              aria-labelledby="login-tab"
            >
              <div class="md-form mb-2 mt-0">
                <md-icon class="prefix grey-text">email</md-icon>
                <input
                  type="email"
                  id="login-email"
                  v-model="email"
                  class="form-control validate mb-2"
                />
                <label data-error="wrong" data-success="right" for="login-email">Your email</label>
              </div>
              <div class="md-form mb-2">
                <md-icon class="prefix grey-text">lock</md-icon>
                <input
                  type="password"
                  id="login-pass"
                  v-model="password"
                  class="form-control validate mb-2"
                  @keyup.enter="login"
                />
                <label data-error="wrong" data-success="right" for="login-pass">Your password</label>
              </div>
              <button class="btn btn-primary btn-block mb-2 mt-4" @click="login">Login</button>
            </div>
            <!-- Register -->
            <div class="tab-pane fade" id="register" role="tabpanel" aria-labelledby="register-tab">
              <div class="md-form pb-2 mt-2">
                <md-icon class="prefix grey-text">account_circle</md-icon>
                <input
                  type="text"
                  id="register-name"
                  v-model="name"
                  class="form-control validate mb-2"
                />
                <label data-error="wrong" data-success="right" for="register-name">Your name</label>
              </div>
              <div class="md-form pb-2 mt-2">
                <md-icon class="prefix grey-text">email</md-icon>
                <input
                  type="email"
                  id="register-email"
                  v-model="email"
                  class="form-control validate mb-2 mt-4"
                />
                <label data-error="wrong" data-success="right" for="register-email">Your email</label>
              </div>
              <div class="md-form pb-2 mt-2">
                <md-icon class="prefix grey-text">lock</md-icon>
                <input
                  type="password"
                  id="register-pass"
                  v-model="password"
                  class="form-control validate mb-2"
                />
                <label data-error="wrong" data-success="right" for="register-pass">Your password</label>
              </div>
              <button class="btn btn-default btn-block mb-2 mt-4" @click="registerUser">Register</button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { fb, db } from "@/firebase";
import { mapGetters, mapMutations } from "vuex";

export default {
  name: "Login",
  data() {
    return {
      name: null,
      email: null,
      password: null,
      userData: null
    };
  },
  computed: {
    ...mapGetters(["user"]),
    nextRoute() {
      return this.$route.query.redirect || "/";
    }
  },
  methods: {
    ...mapMutations(["updateUser"]),
    registerUser() {
      fb.auth()
        .createUserWithEmailAndPassword(this.email, this.password)
        .then(user => {
          db.collection("profiles")
            .doc(user.user.uid)
            .set({
              name: this.name
            });
          $("#userLoginModal").modal("hide");
          this.$router.replace("admin-dashboard");
        })
        .catch(function(error) {
          var errorCode = error.code;
          var errorMessage = error.message;
          if (errorCode == "auth/weak-password") {
            alert("The password is too weak.");
          } else {
            alert(errorMessage);
          }
          console.log(error);
        });
    },
    login() {
      fb.auth()
        .signInWithEmailAndPassword(this.email, this.password)
        .then(() => {
          $("#userLoginModal").modal("hide");
          this.$router.push("/admin-dashboard");
        })
        .catch(function(error) {
          var errorCode = error.code;
          var errorMessage = error.message;
          if (errorCode === "auth/wrong-password") {
            alert("Wrong password.");
          } else {
            alert(errorMessage);
          }
          console.log(error);
        });
    }
  },
  watch: {
    user(auth) {
      if (!!auth) {
        this.$router.replace(this.nextRoute);
      }
    }
  }
};
</script>

<style lang="scss" scoped>
.md-form {
  .prefix {
    top: 0.4rem;
  }
}
.modal-header {
  border-bottom: 0;
  background-color: $main-blue;
}
.modal-footer {
  border-top: 0;
}
.nav-tabs {
  width: 100%;
  border-bottom: 0;

  .nav-link {
    text-transform: uppercase;
    color: #fff;

    &:hover {
      text-decoration: none;
    }

    &:not(.active):hover {
      border-color: transparent;
      background-color: rgba(255, 255, 255, 0.3);
    }
  }

  .nav-item {
    flex: 1 !important;
  }
}
</style>
