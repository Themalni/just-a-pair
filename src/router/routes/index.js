export default [{
    path: '/',
    name: 'Home',
    component: () => import( /* webpackChunkName: "Home" */ '@/views/Home')
},
{
    path: '/product/:id',
    name: 'Product',
    component: () => import( /* webpackChunkName: "Product" */ '@/views/Product')
},
{
    path: '/user-dashboard',
    name: 'UserDashboard',
    component: () => import( /* webpackChunkName: "User Dashboard" */ '@/views/user/UserDashboard'),
    meta: {
        authRequired: true
    }
}, {
    path: '/admin-dashboard',
    name: 'AdminDashboard',
    component: () => import( /* webpackChunkName: "Admin Dashboard" */ '@/views/admin/AdminDashboard'),
    meta: {
        authRequired: true
    }
}
]