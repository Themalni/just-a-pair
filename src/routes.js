import ProductView from './components/ProductView.vue'
import HomePage from './components/HomePage.vue'


export default [
    {path: '/', component: HomePage},
    {path: '/product-view/:id', component: ProductView }
]