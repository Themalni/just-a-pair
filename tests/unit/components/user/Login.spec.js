import { mount, createLocalVue } from '@vue/test-utils';
import Login from '@/components/user/Login';
import mockState from '../../utils/mockState';
import Vuex from 'vuex';
import VueRouter from "vue-router";
import routes from "@/router/routes";
import $ from 'jquery';

global.$ = $;

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(VueRouter);

describe('Login', () => {
    let wrapper
    let store
    let state
    let router

    beforeEach(() => {
        router = new VueRouter({ routes })
        state = mockState.state
        store = new Vuex.Store({
            state: state
        })

        wrapper = mount(Login, {
            stubs: ['md-icon'],
            store,
            localVue,
            router
        });
    });

    afterEach(() => {
        wrapper.destroy();
    });

    it('renders correctrly', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('registers new user', async () => {
        wrapper.vm.email = 'john_doe@gmail.com';
        wrapper.vm.password = '12Dfrt23#';

        const registerUserMock = jest.spyOn(wrapper.vm, 'registerUser');

        await wrapper.find('.register-btn').trigger('click');

        expect(registerUserMock).toBeCalled();
    });
})