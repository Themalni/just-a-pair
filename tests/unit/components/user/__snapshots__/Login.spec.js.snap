// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders correctrly 1`] = `
<div id="userLoginModal" tabindex="-1" role="dialog" aria-hidden="true" class="modal fade">
  <div role="document" class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header text-center">
        <ul id="loginTab" role="tablist" class="nav nav-tabs">
          <li class="nav-item"><a id="login-tab" data-toggle="tab" href="#login" role="tab" aria-controls="login" aria-selected="true" class="nav-link active mr-1">Login</a></li>
          <li class="nav-item"><a id="register-tab" data-toggle="tab" href="#register" role="tab" aria-controls="register" aria-selected="false" class="nav-link ml-1">Register</a></li>
        </ul>
      </div>
      <div class="modal-body mx-3">
        <div id="myTabContent" class="tab-content">
          <div id="login" role="tabpanel" aria-labelledby="login-tab" class="tab-pane fade show active">
            <div class="md-form mb-2 mt-3">
              <md-icon-stub class="prefix grey-text">email</md-icon-stub> <input type="email" id="login-email" class="form-control validate mb-2"> <label data-error="wrong" data-success="right" for="login-email">Your email</label>
            </div>
            <div class="md-form mb-2">
              <md-icon-stub class="prefix grey-text">lock</md-icon-stub> <input type="password" id="login-pass" class="form-control validate mb-2"> <label data-error="wrong" data-success="right" for="login-pass">Your password</label>
            </div> <button class="login-btn btn btn-primary btn-block mb-2 mt-4">Login</button>
          </div>
          <div id="register" role="tabpanel" aria-labelledby="register-tab" class="tab-pane fade">
            <div class="md-form pb-2 mt-2">
              <md-icon-stub class="prefix grey-text">account_circle</md-icon-stub> <input type="text" id="register-name" class="form-control validate mb-2"> <label data-error="wrong" data-success="right" for="register-name">Your name</label>
            </div>
            <div class="md-form pb-2 mt-2">
              <md-icon-stub class="prefix grey-text">email</md-icon-stub> <input type="email" id="register-email" class="form-control validate mb-2 mt-4"> <label data-error="wrong" data-success="right" for="register-email">Your email</label>
            </div>
            <div class="md-form pb-2 mt-2">
              <md-icon-stub class="prefix grey-text">lock</md-icon-stub> <input type="password" id="register-pass" class="form-control validate mb-2"> <label data-error="wrong" data-success="right" for="register-pass">Your password</label>
            </div> <button class="register-btn btn btn-default btn-block mb-2 mt-4">Register</button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
